# US Immigration Data with Detailed Profile of Entry/Destination States and Cities
## Data Engineering Capstone Project

### Project Summary
The purpose of this project is to gather and transform US immigration data and its related data including US demographic data, temperature data, and crime data, so that one can obtain more detailed analysis results on the US immigration data.

The project follows the follow steps:
* Step 1: Scope the Project and Gather Data
* Step 2: Explore and Assess the Data
* Step 3: Define the Data Model
* Step 4: Run ETL to Model the Data
* Step 5: Complete Project Write Up

In [1]:
# Do all imports and installs here
import pandas as pd
import re
import os
import glob
import psycopg2
import pandas as pd
from sql_queries import *
from pyspark.sql import SparkSession
from datetime import datetime
from pyspark.sql.types import *
from pyspark.sql import types as T
from pyspark.sql import functions as F
from pyspark.sql import SQLContext
from pyspark.sql.functions import *
from datetime import datetime, timedelta

## Step 1: Scope the Project and Gather Data

### Scope 
#### Solution goal:
The goal is to create a database schema that consists of 1 fact table and 4 dimension tables.  
- A **fact** table would be **immigration** table, created by combining Immigration data with different data sources whose primary key would be *id*.
- The first **dimension** table would be **demographic** table, whose primary key would be *state*.
- The second **dimension** table would be **crime** table, whose composite key would be *county_name* and *state_name*.
- The third **dimension** table would be **temperature** table, whose composite key would be *county* and *state*.
- The last **dimension** table would be **date** table, whose primary key would be *created_date*.


#### Data to use:
- `I94 Immigration Data` - US Immigration data of 2016 from "US National Tourism and Trade Office". It includes information about immigrants such as origin city, reason for immigration, visa type, gender, birth year, flight number and so on. 
- `I94_SAS_Labels_Descriptions.SAS` - Text file which includes lists of codes and corresponding lables for the `I94 Immigration Data`.
- `us-cities-demographics.csv`   - Basic demographic information of the US of 2015 which comes from "OpenSoft". It includes statistics by state, city, race, age, gender and so on.
- `airport-codes.csv.csv`  -  List of all airport codes which comes from "DATA HUB". It includes airport information such as airpot type, name, municipality and so on.
- `Offenses_Known_to_Law_Enforcement.xls`   - Number of crimes by US state and metropolitan and non-metropolitan country of 2015 which comes from "data.world". It includes number of crimes in each crime category in the US.
- `GlobalLandTemperaturesByCity.csv` - Global temperature information since 1970s by country and cities which come from "Kaggle". It includes monthly average temperatures.
- `uscities.csv` - Basic information about US cities, counties, and states.


#### Tool to use:
- `Jupyter Notebook`
- `Python`
- `Pandas`
- `PySpark`
- `PostgreSQL`

In [2]:
# Creating spark cluster
from pyspark.sql import SparkSession
spark = SparkSession.builder.\
config("spark.jars.packages","saurfang:spark-sas7bdat:2.0.0-s_2.11")\
.enableHiveSupport().getOrCreate()

In [3]:
# Reading in the immigration data
immigration_data=spark.read.parquet("sas_data")

In [4]:
# Checking number of records in the immigration data
immigration_data.count()

3096313

In [5]:
# Read in the sample data of the immigration data and show the first few rows to see what the contents look like.
fname = 'immigration_data_sample.csv'
df_imgr = pd.read_csv(fname)
df_imgr.head()

Unnamed: 0      cicid   i94yr  i94mon  i94cit  i94res i94port  arrdate  \
0     2027561  4084316.0  2016.0     4.0   209.0   209.0     HHW  20566.0   
1     2171295  4422636.0  2016.0     4.0   582.0   582.0     MCA  20567.0   
2      589494  1195600.0  2016.0     4.0   148.0   112.0     OGG  20551.0   
3     2631158  5291768.0  2016.0     4.0   297.0   297.0     LOS  20572.0   
4     3032257   985523.0  2016.0     4.0   111.0   111.0     CHM  20550.0   

   i94mode i94addr    ...     entdepu  matflag  biryear   dtaddto  gender  \
0      1.0      HI    ...         NaN        M   1955.0  07202016       F   
1      1.0      TX    ...         NaN        M   1990.0  10222016       M   
2      1.0      FL    ...         NaN        M   1940.0  07052016       M   
3      1.0      CA    ...         NaN        M   1991.0  10272016       M   
4      3.0      NY    ...         NaN        M   1997.0  07042016       F   

  insnum airline        admnum  fltno  visatype  
0    NaN      JL  5.658267e+10  00782        WT  
1    NaN     *GA  9.436200e+10  XBLNG        B2  
2    NaN      LH  5.578047e+10  00464        WT  
3    NaN      QR  9.478970e+10  00739        B2  
4    NaN     NaN  4.232257e+10   LAND        WT  

[5 rows x 29 columns]

In [6]:
# Read in the crime data
df_crime = pd.read_excel('Offenses_Known_to_Law_Enforcement.xls', 
                         header=4, 
                         names=["State",
                                "County",
                                "Violent_crime",
                                "Murder_and_nonnegligent_manslaughter",
                                "Rape_1",
                                "Rape_2",
                                "Robbery",
                                "Aggravated_assault",
                                "Property_crime",
                                "Burglary",
                                "Larceny-theft",
                                "Motor_vehicle_theft",
                                "Arson3"])
df_crime.head(5)

State   County  Violent_crime  \
0  ALABAMA - Metropolitan Counties  Autauga             69   
1  ALABAMA - Metropolitan Counties  Baldwin            115   
2  ALABAMA - Metropolitan Counties     Bibb              7   
3  ALABAMA - Metropolitan Counties   Blount            204   
4  ALABAMA - Metropolitan Counties  Calhoun             16   

   Murder_and_nonnegligent_manslaughter  Rape_1  Rape_2  Robbery  \
0                                     0    13.0     NaN        6   
1                                     0     9.0     NaN       33   
2                                     0     2.0     NaN        1   
3                                     5    16.0     NaN        5   
4                                     0     4.0     NaN        1   

   Aggravated_assault  Property_crime  Burglary  Larceny-theft  \
0                  50           344.0     111.0          187.0   
1                  73           648.0     225.0          390.0   
2                   4            41.0      20.0           18.0   
3                 178           832.0     247.0          503.0   
4                  11           413.0     181.0          225.0   

   Motor_vehicle_theft  Arson3  
0                 46.0     NaN  
1                 33.0     NaN  
2                  3.0     NaN  
3                 82.0     NaN  
4                  7.0     NaN

In [7]:
# Read in the us demographic data
df_demo = pd.read_csv('us-cities-demographics.csv', header=None)


# Split a column which has multiple values into separate columns and set it as a dataframe
df_demo = df_demo[0].str.split(";",expand = True)
df_demo = pd.DataFrame(df_demo)

# Change headers
headers = df_demo.iloc[0]
df_demo  = pd.DataFrame(df_demo.values[1:], columns=headers)

df_demo.head(5)

0              City          State Median Age Male Population  \
0     Silver Spring       Maryland       33.8           40601   
1            Quincy  Massachusetts       41.0           44129   
2            Hoover        Alabama       38.5           38040   
3  Rancho Cucamonga     California       34.5           88127   
4            Newark     New Jersey       34.6          138040   

0 Female Population Total Population Number of Veterans Foreign-born  \
0             41862            82463               1562        30908   
1             49500            93629               4147        32935   
2             46799            84839               4819         8229   
3             87105           175232               5821        33878   
4            143873           281913               5829        86253   

0 Average Household Size State Code                       Race  Count  
0                    2.6         MD         Hispanic or Latino  25924  
1                   2.39         MA                      White  58723  
2                   2.58         AL                      Asian   4759  
3                   3.18         CA  Black or African-American  24437  
4                   2.73         NJ                      White  76402

In [8]:
# Read in Temperature data
df_temp = pd.read_csv('../../data2/GlobalLandTemperaturesByCity.csv')
df_temp.head(5)

dt  AverageTemperature  AverageTemperatureUncertainty   City  \
0  1743-11-01               6.068                          1.737  Århus   
1  1743-12-01                 NaN                            NaN  Århus   
2  1744-01-01                 NaN                            NaN  Århus   
3  1744-02-01                 NaN                            NaN  Århus   
4  1744-03-01                 NaN                            NaN  Århus   

   Country Latitude Longitude  
0  Denmark   57.05N    10.33E  
1  Denmark   57.05N    10.33E  
2  Denmark   57.05N    10.33E  
3  Denmark   57.05N    10.33E  
4  Denmark   57.05N    10.33E

In [9]:
# Read in US cities data
df_label = pd.read_csv('uscities.csv')
df_label.head(5)

city   city_ascii state_id  state_name  county_fips  county_name  \
0     New York     New York       NY    New York        36061     New York   
1  Los Angeles  Los Angeles       CA  California         6037  Los Angeles   
2      Chicago      Chicago       IL    Illinois        17031         Cook   
3        Miami        Miami       FL     Florida        12086   Miami-Dade   
4       Dallas       Dallas       TX       Texas        48113       Dallas   

       lat       lng  population  density   source  military  incorporated  \
0  40.6943  -73.9249    18713220    10715  polygon     False          True   
1  34.1139 -118.4068    12750807     3276  polygon     False          True   
2  41.8373  -87.6862     8604203     4574  polygon     False          True   
3  25.7839  -80.2102     6445545     5019  polygon     False          True   
4  32.7936  -96.7662     5743938     1526  polygon     False          True   

              timezone  ranking  \
0     America/New_York        1   
1  America/Los_Angeles        1   
2      America/Chicago        1   
3     America/New_York        1   
4      America/Chicago        1   

                                                zips          id  
0  11229 11226 11225 11224 11222 11221 11220 1138...  1840034016  
1  90291 90293 90292 91316 91311 90037 90031 9000...  1840020491  
2  60018 60649 60641 60640 60643 60642 60645 6064...  1840000494  
3  33129 33125 33126 33127 33128 33149 33144 3314...  1840015149  
4  75287 75098 75233 75254 75251 75252 75253 7503...  1840019440

In [10]:
# Read in airport data
df_airport = pd.read_csv('airport-codes_csv.csv')
df_airport = df_airport.rename(
    columns={'name': 'airport_name',
             'iso_region':'entry_state_id',
             'municipality':'entry_county'})
df_airport = df_airport[df_airport['entry_state_id'].str.contains('US-')]
df_airport['entry_state_id'] = [x.replace('US-','') for x in df_airport['entry_state_id']]
df_airport = df_airport[['local_code',
                         'airport_name',
                         'entry_state_id',
                         'entry_county']]
df_airport.head(5)

local_code                        airport_name entry_state_id  entry_county
0        00A                   Total Rf Heliport             PA      Bensalem
1       00AA                Aero B Ranch Airport             KS         Leoti
2       00AK                        Lowell Field             AK  Anchor Point
3       00AL                        Epps Airpark             AL       Harvest
4        NaN  Newport Hospital & Clinic Heliport             AR       Newport

### Step 2: Explore and Assess the Data
#### Explore the Data 
Identify data quality issues, like missing values, duplicate data, etc.


#### Cleaning Steps

- US Demographic data
   (Data: `us-cities-demographics.csv`)
  1. Change data types of each column
  2. Format columns to keep consistensy among other tables
  3. Break down 'race' column into separate columns for each racial groups and sum up the numbers by each state
  4. Select and sort columns to finalize the Demographic table


- Crime data
   (Data: `Offenses_Known_to_Law_Enforcement.xls`, `uscities.csv`)
  1. Split a column which has more than 1 elements into separate columns
  2. Change data types of each column
  3. Sum up some of the column values to simplify the table. Also get a total value for all crime numbers
  4. Join with `uscities.csv` table to acquire total population of each county
  5. Calculate crime rate by county
  6. Select, sort and rename columns to finalize table


- Immigiration data
    (Data: `I94 Immigration Data`, `I94_SAS_Labels_Descriptions.SAS`, `uscities.csv`, - `airport-codes.csv.csv` )
  1. For those columns that have many missing values or that are not important for this project, deleted them, change column data types, and rename column names to make it more understandable
  2. In order to add label columns from other tables to Immigration table for acquiring code lables, format `I94_SAS_Labels_Descriptions.SAS` and `uscities.csv` file to join
  3. Create and format label tables for ports, countries, and visa
  4. Format `uscities.csv` table to join state names
  5. Convert pandas dataframes to pyspark dataframes for joining with `I94 Immigration Data` table
  6. Join converted dataframes with `I94 Immigration Data` table to add label columns
  7. Select and sort columns, filter to finalize the Immigiration table


- Temperature data
    (Data: `GlobalLandTemperaturesByCity.csv`, `uscities.csv`)
  1. Format `uscities.csv` table to join with `GlobalLandTemperaturesByCity.csv` table
  2. Delete rows with null values, filter with only US data
  3. Aggregate table to get avg, min, and max value of temperature for each city
  4. Join with uscities table to add county column
  5. Select and sort columns to finalize the Temperature table

In [11]:
# Cleaning US Demographic table

# Change data type of each column
cols = ['Male Population',
        'Female Population',
        'Number of Veterans',
       'Foreign-born',
        'Average Household Size',
        'Count']
df_demo[cols] = df_demo[cols].apply(pd.to_numeric, 
                                    errors='coerce', 
                                    axis=1).fillna(0.0)
df_demo = df_demo.astype({'Median Age': float, 
                          'Male Population': int, 
                          'Female Population': int,
                          'Number of Veterans':int, 
                           'Foreign-born':int, 
                          'Average Household Size':int, 
                          'Count':int})

# Format a column to make it upper case to keep consistensy among other tables
df_demo['State'] = df_demo['State'].str.upper()

# Break down 'race' column into separate columns for each racial groups and get the sum numbers of each state
df_demo_race = df_demo.groupby(['State','Race'])\
.agg({'Count':'sum'})\
.unstack(fill_value=0)\
.reset_index()
df_demo_race.columns = df_demo_race.columns.droplevel(0)
df_demo_race = df_demo_race.rename(columns={"":'State'})
df_demo = df_demo.groupby(['State']).agg({'State Code':'max',
                                          'Count':'sum',
                                          'Median Age':'median',
                                          'Male Population':'sum',
                                         'Female Population':'sum',
                                          'Number of Veterans':'sum',
                                          'Foreign-born':'sum',
                                         'Average Household Size':'mean'})
df_demo['Average Household Size'] = df_demo['Average Household Size'].round(2)
df_demo = pd.merge(df_demo, df_demo_race, left_on='State', right_on='State')


# Select and sort columns
demographic = df_demo[['State','State Code',
                       'Count',
                       'Median Age',
                       'Male Population',
                       'Female Population',
                       'Number of Veterans',
                       'Foreign-born',
                       'Average Household Size',
                       'American Indian and Alaska Native',
                       'Asian','Black or African-American',
                       'Hispanic or Latino',
                       'White']].sort_values('State')
demographic = demographic.rename(columns={'State': 'state', 
                                          'State Code':'state_code',
                                          'Count':'population',
                                          'Median Age':'median_age',
                                          'Male Population':'male_population',
                                          'Female Population':'female_population',
                                          'Number of Veterans':'number_of_veterans',
                                          'Foreign-born':'foreign_born',
                                          'Average Household Size':'average_household_size',
                                          'American Indian and Alaska Native':'AmericanIndian_and_AlaskaNative_population',
                                          'Asian':'Asian_population',
                                          'Black or African-American':'Black_or_AfricanAmerican_population',
                                          'Hispanic or Latino':'Hispanic_or_Latino_population',
                                          'White':'White_population'})
# Replace null value by 0
demographic = demographic.fillna(0)

demographic.head()

state state_code  population  median_age  male_population  \
0     ALABAMA         AL     1096154        38.0          2448200   
1      ALASKA         AK      336228        32.2           764725   
2     ARIZONA         AZ     5754881        34.1         11137275   
3    ARKANSAS         AR      643597        32.6          1400724   
4  CALIFORNIA         CA    31753718        35.8         61055672   

   female_population  number_of_veterans  foreign_born  \
0            2715106              352896        252541   
1             728750              137460        166290   
2           11360435             1322525       3411565   
3            1482165              154390        307753   
4           62388681             4617022      37059662   

   average_household_size  AmericanIndian_and_AlaskaNative_population  \
0                    2.00                                        8084   
1                    2.00                                       36339   
2                    2.19                                      129708   
3                    2.17                                        9381   
4                    2.59                                      401386   

   Asian_population  Black_or_AfricanAmerican_population  \
0             28769                               521068   
1             36825                                23107   
2            229183                               296222   
3             22062                               149608   
4           4543730                              2047009   

   Hispanic_or_Latino_population  White_population  
0                          39313            498920  
1                          27261            212696  
2                        1508157           3591611  
3                          77813            384733  
4                        9856464          14905129

In [12]:
# Cleaning and aggregating Crime table

# Split a column which has more than 1 elements into separate columns
state = df_crime["State"].str.split("-",expand = True) 
df_crime["State"] = state[0]
df_crime["category"] = state[1]

# Change data types of each column
df_crime = df_crime.fillna(0)
df_crime = df_crime.astype({'Rape_1': int, 
                            'Rape_2': int, 
                            'Property_crime': int,
                            'Burglary':int, 
                            'Larceny-theft':int, 
                            'Motor_vehicle_theft':int, 
                            'Arson3':int})

# Sum up some of the column values to simplify the table. Also get a total value for all crime numbers
df_crime['Rape'] = df_crime['Rape_1'] + df_crime['Rape_2']
df_crime = df_crime.drop(['Rape_1', 
                          'Rape_2'], axis=1)
df_crime['total_crime'] = df_crime['Violent_crime']\
+ df_crime['Murder_and_nonnegligent_manslaughter']\
+ df_crime['Robbery']\
+ df_crime['Burglary']\
+ df_crime['Larceny-theft']\
+ df_crime['Motor_vehicle_theft']\
+ df_crime['Arson3']

# Cleaning County names and sum up
df_crime['State'] = df_crime['State'].str.strip()
df_crime['County'] = [re.sub(r'\d+|\,','', x) for x in df_crime['County']]
df_crime['County_fixed'] = df_crime['County']\
.replace('Police| Department| Unified| County| Public Safety', "", regex=True)
df_crime['County_fixed'] = df_crime['County_fixed'].str.strip()

# Join US Demographic table and Crime table to get total population of states.
df_label['county_new'] = df_label['county_name']\
.mask((df_label['county_name']=='LaSalle') & (df_label['state_name']=='Illinois'), 
      'La Salle')\
.mask((df_label['county_name']=='LaPorte') & (df_label['state_name']=='Indiana'), 
      'La Porte')\
.mask((df_label['county_name']=='DeKalb') & (df_label['state_name']=='Indiana'), 
      'De Kalb')\
.mask((df_label['county_name']=='LaSalle') & (df_label['state_name']=='Louisiana'), 
      'La Salle')\
.mask((df_label['county_name']=='Lac qui Parle') & (df_label['state_name']=='Minnesota'), 
      'Lac Qui Parle')\
.mask((df_label['county_name']=='Trousdale') & (df_label['state_name']=='Tennessee'), 
      'Hartsville/Trousdale')\
.mask((df_label['county_name']=='Doña Ana') & (df_label['state_name']=='New Mexico'), 
      'Dona Ana')\
.mask((df_label['county_name']=='Dutchess') & (df_label['state_name']=='New York'), 
      'Duchess')\
.mask((df_label['county_name']=='LaMoure') & (df_label['state_name']=='North Dakota'), 
      'Lamoure')
df_label_for_join = df_label[['state_name',
                              'county_new',
                              'population']]
df_label_for_join = pd.DataFrame(df_label_for_join.groupby(['state_name',
                                                            'county_new'])\
                                 .agg({'population':'sum'})).reset_index()
df_label_for_join['state_name'] = df_label_for_join['state_name'].str.upper()

# Join population column from demographic table to crime table and calculate crime rate
crime = pd.merge(df_crime, 
                 df_label_for_join, 
                 left_on=['State','County_fixed'], 
                 right_on=['state_name','county_new'], 
                 how='inner')
crime['crime_rate'] = (crime['total_crime']/crime['population']*100000)\
.astype(float).round(1)
crime['violent_crime_rate'] = (crime['Violent_crime']/crime['population']*100000)\
.astype(float).round(1)
crime['property_crime_rate'] = (crime['Property_crime']/crime['population']*100000)\
.astype(float).round(1)
crime['population'] = crime['population'].astype(int)
crime = crime[['County_fixed',
               'State','category',
               'total_crime',
               'Violent_crime',
               'Murder_and_nonnegligent_manslaughter',
               'Robbery','Aggravated_assault',
               'Property_crime',
               'Burglary',
               'Larceny-theft',
               'Motor_vehicle_theft',
               'Arson3',
               'Rape',
               'crime_rate',
               'violent_crime_rate',
               'property_crime_rate']]
crime = crime.rename(columns={'County_fixed':'county',
                              'State':'state',
                              'Violent_crime':'violent_crime',
                              'Murder_and_nonnegligent_manslaughter':'murder_and_nonnegligent_manslaughter',
                              'Robbery':'robbery',
                              'Aggravated_assault':'aggravated_assault',
                              'Property_crime':'property_crime',
                              'Burglary':'burglary',
                              'Larceny-theft':'larceny-theft',
                              'Motor_vehicle_theft':'motor_vehicle_theft',
                              'Arson3':'arson3',
                              'Rape':'rape'})
crime.head(5)

county    state                category  total_crime  violent_crime  \
0  Autauga  ALABAMA   Metropolitan Counties          419             69   
1  Baldwin  ALABAMA   Metropolitan Counties          796            115   
2     Bibb  ALABAMA   Metropolitan Counties           49              7   
3   Blount  ALABAMA   Metropolitan Counties         1046            204   
4  Calhoun  ALABAMA   Metropolitan Counties          430             16   

   murder_and_nonnegligent_manslaughter  robbery  aggravated_assault  \
0                                     0        6                  50   
1                                     0       33                  73   
2                                     0        1                   4   
3                                     5        5                 178   
4                                     0        1                  11   

   property_crime  burglary  larceny-theft  motor_vehicle_theft  arson3  rape  \
0             344       111            187                   46       0    13   
1             648       225            390                   33       0     9   
2              41        20             18                    3       0     2   
3             832       247            503                   82       0    16   
4             413       181            225                    7       0     4   

   crime_rate  violent_crime_rate  property_crime_rate  
0       972.5               160.1                798.4  
1       406.1                58.7                330.6  
2       366.6                52.4                306.8  
3      5878.7              1146.5               4676.0  
4       309.1                11.5                296.9

In [13]:
# Check number of null values in each column for data cleaning
#immigration_data.select([count(when(col(c).isNull(), c)).alias(c) for c in immigration_data.columns]).show()

In [14]:
# Cleaning Immigration table

def convert_datetime(x):
    """a function to convert integer to date format"""
    try:
        start = datetime(1960, 1, 1)
        return start + timedelta(days=int(x))
    except:
        return None
udf_datetime_from_sas = udf(lambda x: convert_datetime(x), T.DateType())

# For those columns that have many missing values or that are not important for this project, deleted them, 
# change column data types, and rename column names to make it more understandable
immigration_data = immigration_data.drop("i94mode")\
.drop('occup')\
.drop('entdepu')\
.drop("insnum")\
.drop('entdepa')\
.drop('matflag')\
.drop('count')\
.drop('visatype')\
.drop('visapost')\
.drop('entdepd')\
.drop('year')\
.drop('month')\
.withColumn("id", col("cicid").cast("integer")) \
.drop("cicid") \
.withColumn("year", col("i94yr").cast("integer")) \
.drop("i94yr") \
.withColumn("month", col("i94mon").cast("integer")) \
.drop("i94mon") \
.withColumn("origin_country", col("i94cit").cast("bigint")) \
.drop("i94cit") \
.withColumn("resident_country", col("i94res").cast("bigint")) \
.drop("i94res") \
.withColumnRenamed("i94port", "port_of_entry")\
.withColumnRenamed("i94addr", "state_code")\
.withColumn("birth_year", col("biryear").cast("integer")) \
.drop("biryear") \
.withColumn("visa_type", col("i94visa").cast("bigint")) \
.drop("i94visa") \
.withColumn("age", col("i94bir").cast("integer")) \
.drop("i94bir") \
.withColumn("admission_no", col("admnum").cast("integer")) \
.drop("admnum") \
.withColumnRenamed("fltno", "flight_no")\
.withColumn('admitted_date',F.to_date(F.unix_timestamp(immigration_data.dtaddto, 'MMddyyyy').cast('timestamp')))\
.drop("dtaddto")\
.withColumn('created_date',F.to_date(F.unix_timestamp(immigration_data.dtadfile, 'yyyyMMdd').cast('timestamp')))\
.drop("dtadfile")\
.withColumn("departure_date", udf_datetime_from_sas("depdate"))\
.drop("depdate")\
.withColumn("arrival_date", udf_datetime_from_sas("arrdate"))\
.drop("arrdate")

# Delete null values
immigration_data = immigration_data.dropna("any")

In [15]:
# In order to add label columns from other tables to Immigration table for acquiring code lables, format I94_SAS_Labels_Descriptions.SAS and uscities.csv file to join
sqlContext = SQLContext(spark)

def equivalent_type(f):
    if f == 'datetime64[ns]': return TimestampType()
    elif f == 'int64': return LongType()
    elif f == 'int32': return IntegerType()
    elif f == 'float64': return FloatType()
    else: return StringType()

def define_structure(string, format_type):
    try: typo = equivalent_type(format_type)
    except: typo = StringType()
    return StructField(string, typo)

def pandas_to_spark(pandas_df):
    columns = list(pandas_df.columns)
    types = list(pandas_df.dtypes)
    struct_list = []
    for column, typo in zip(columns, types): 
        struct_list.append(define_structure(column, typo))
    p_schema = StructType(struct_list)
    return sqlContext.createDataFrame(pandas_df, p_schema)

# Open and read a text file
with open("./I94_SAS_Labels_Descriptions.SAS") as f:
    textfile = f.readlines()
    
# Create and format label tables for ports, countries, and visa
def process_text_file(file):
    """Convert text file into a dataframe format"""
    df = list(map(lambda x:x.strip(),file))
    df = [re.sub(r'\t','', x) for x in df]
    df = [x.replace('=',',') for x in df]
    df = [x.replace('\'','') for x in df]
    df = pd.DataFrame(df)
    df = df[0].str.split(",",expand = True)
    df[0] = df[0].str.strip()
    df[1] = df[1].str.strip()
    return df[[0,1]]

ports = textfile[302:962]
df_ports = process_text_file(ports)
df_ports = df_ports.rename(columns={0: 'entry_port_code', 
                                    1: 'port_name'})

countries = textfile[10:298]
df_countries = process_text_file(countries)
df_countries[0] = df_countries[0].astype(int)
df_countries_origin = df_countries.rename(columns={0: 'country_code', 
                                                   1: 'origin_country_name'})
df_countries_resident = df_countries.rename(columns={0: 'country_code', 
                                                     1: 'resident_country_name'})

visa = textfile[1046:1049]
df_visa = process_text_file(visa)
df_visa[0] = df_visa[0].astype(int)
df_visa = df_visa.rename(columns={0: 'visa_type_code', 
                                  1: 'visa_type_name'})

# Format df_label table to join state names
df_label_state = df_label[['state_name',
                           'city',
                           'state_id',
                           'county_name']]
df_label_state = pd.DataFrame(df_label_state
                              .groupby(['state_id'])
                              .agg({'state_name':'max'})).reset_index()
df_label_state['state_name'] = df_label_state['state_name'].str.upper().str.strip()
df_label_current = df_label_state.rename(columns={'state_name': 'current_state_name'})
df_label_entry = df_label_state.rename(columns={'state_name': 'entry_state_name'})

# Convert pandas dataframes to pyspark dataframes for joining with Immigration table
dfs = [df_ports, 
       df_countries_origin, 
       df_countries_resident, 
       df_visa, df_airport, 
       df_label_current, 
       df_label_entry]
dfs_new = [pandas_to_spark(file) for file in dfs]
df_ports_for_join, \
df_countries_origin_for_join, \
df_countries_resident_for_join, \
df_visa_for_join, \
df_airport_for_join, \
df_label_current_for_join, \
df_label_entry_for_join = dfs_new

# Join converted dataframes with Immigration table to add columns
immigration_data = immigration_data.join(df_ports_for_join, 
                                         immigration_data.port_of_entry == df_ports_for_join.entry_port_code, 
                                         "left")
immigration_data = immigration_data.join(df_countries_origin_for_join, 
                                         immigration_data.origin_country == df_countries_origin_for_join.country_code, 
                                         "left")
immigration_data = immigration_data.join(df_countries_resident_for_join, 
                                         immigration_data.resident_country == df_countries_resident_for_join.country_code, 
                                         "left")
immigration_data = immigration_data.join(df_visa_for_join, 
                                         immigration_data.visa_type == df_visa_for_join.visa_type_code, 
                                         "left")
immigration_data = immigration_data.join(df_airport_for_join, 
                                         immigration_data.port_of_entry == df_airport_for_join.local_code, 
                                         "left")
immigration_data = immigration_data.join(df_label_current_for_join, 
                                         immigration_data.state_code == df_label_current_for_join.state_id, 
                                         "left")
immigration_data = immigration_data.join(df_label_entry_for_join, 
                                         immigration_data.entry_state_id == df_label_entry_for_join.state_id, 
                                         "left")

# Select and sort columns
immigration_data = immigration_data.drop("country_code")\
.drop('origin_country')\
.drop('resident_country')\
.drop("entry_port_code")\
.drop("visa_type_code")\
.drop("local_code")\
.drop("state_id")

immigration = immigration_data.select("id",
                                      "created_date",
                                      "admission_no",
                                      "admitted_date",
                                      "departure_date",
                                      "arrival_date",
                                      "port_name",
                                      "airport_name",
                                      "entry_state_id",
                                      "entry_state_name",
                                      "entry_county",
                                      "airline",
                                      "flight_no",
                                      "gender",
                                      "birth_year",
                                      "age",
                                      "origin_country_name",
                                      "resident_country_name",
                                      "visa_type",
                                      "visa_type_name",
                                      "state_code",
                                      'current_state_name')

# Shrink data into sample rows for the sake of a learning project -- this row should be filtered out in actual environment
immigration = spark.createDataFrame(immigration.sample(0.01).collect())

immigration.head(5)

[Row(id=1973857, created_date=datetime.date(2016, 4, 11), admission_no=2147483647, admitted_date=datetime.date(2016, 7, 9), departure_date=datetime.date(2016, 4, 25), arrival_date=datetime.date(2016, 4, 11), port_name='NEWARK/TETERBORO', airport_name='Lakefront Airport', entry_state_id='LA', entry_state_name='LOUISIANA', entry_county='New Orleans', airline='UA', flight_no='00998', gender='F', birth_year=1990, age=26, origin_country_name='BELGIUM', resident_country_name='BELGIUM', visa_type=2, visa_type_name='Pleasure', state_code='AZ', current_state_name='ARIZONA'),
 Row(id=3700217, created_date=datetime.date(2016, 4, 20), admission_no=2147483647, admitted_date=datetime.date(2016, 10, 19), departure_date=datetime.date(2016, 5, 22), arrival_date=datetime.date(2016, 4, 20), port_name='NEWARK/TETERBORO', airport_name='Lakefront Airport', entry_state_id='LA', entry_state_name='LOUISIANA', entry_county='New Orleans', airline='UA', flight_no='00961', gender='M', birth_year=1981, age=35, orig

In [16]:
# Creating Date table using Immigration table

date = immigration_data.withColumn("day", dayofmonth("created_date")) \
        .withColumn("week", weekofyear("created_date")) \
        .withColumn("month", month("created_date")) \
        .withColumn("year", year("created_date")) \
        .withColumn("weekday", dayofweek("created_date"))\
    .select("created_date", 
            "day", 
            "week", 
            "month", 
            "year", 
            "weekday").drop_duplicates()
date.show(5)

+------------+---+----+-----+----+-------+
|created_date|day|week|month|year|weekday|
+------------+---+----+-----+----+-------+
|  2016-06-23| 23|  25|    6|2016|      5|
|  2016-04-14| 14|  15|    4|2016|      5|
|  2016-05-24| 24|  21|    5|2016|      3|
|  2016-05-26| 26|  21|    5|2016|      5|
|  2016-05-03|  3|  18|    5|2016|      3|
+------------+---+----+-----+----+-------+
only showing top 5 rows



In [17]:
# Cleaning and aggregating Temperature table

# Format df_label table to join with temperature table
df_county = pd.DataFrame(df_label
                         .groupby(['city'])
                         .agg({'county_name':'max',
                               'state_name':'max'})).reset_index()

# Delete rows with null values, filter with only US data
df_temp = df_temp.dropna()
df_temp['dt'] = pd.to_datetime(df_temp['dt'])
df_temp = df_temp[df_temp['Country']=='United States']

# Aggregate table to get avg, min, and max value of temperature for each city
df_temp_mean = pd.DataFrame(df_temp
                            .groupby(['City'])
                            .agg({'AverageTemperature':'mean'})).reset_index()
df_temp_mean = df_temp_mean.rename(columns = {'AverageTemperature':'avg_temperature'})
df_temp_mean['avg_temperature'] = df_temp_mean['avg_temperature'].round(2)
df_temp_max = pd.DataFrame(df_temp
                           .groupby(['City'])
                           .agg({'AverageTemperature':'max'})).reset_index()
df_temp_max = df_temp_max.rename(columns = {'AverageTemperature':'max_of_monthly_avg_temperature'})
df_temp_max['max_of_monthly_avg_temperature'] = df_temp_max['max_of_monthly_avg_temperature'].round(2)
df_temp_min = pd.DataFrame(df_temp
                           .groupby(['City'])
                           .agg({'AverageTemperature':'min'})).reset_index()
df_temp_min = df_temp_min.rename(columns = {'AverageTemperature':'min_of_monthly_avg_temperature'})
df_temp_min['min_of_monthly_avg_temperature'] = df_temp_min['min_of_monthly_avg_temperature'].round(2)
mean_max = pd.merge(df_temp_mean,df_temp_max)
mean_max_min = pd.merge(mean_max,df_temp_min)

# Join with df_county table to add county column
temp = pd.merge(mean_max_min, 
                df_county, 
                left_on='City', 
                right_on='city', 
                how='inner')
temp = temp.groupby(['county_name',
                     'state_name'])\
.agg({'avg_temperature':'mean',
      'max_of_monthly_avg_temperature':'max',
      'min_of_monthly_avg_temperature':'min'}).reset_index()

# Select and sort columns
temperature = temp[['county_name', 
                    'state_name',
                    'avg_temperature', 
                    'max_of_monthly_avg_temperature', 
                    'min_of_monthly_avg_temperature']].sort_values('county_name')
temperature['state_name'] = temperature['state_name'].str.upper()
temperature = temperature.rename(columns={'county_name':'county',
                                          'state_name':'state'})
temperature.head(5)

county         state  avg_temperature  max_of_monthly_avg_temperature  \
0   Allegheny  PENNSYLVANIA             9.61                           28.60   
1       Allen       INDIANA             9.70                           29.04   
2        Bell         TEXAS            18.54                           32.66   
3  Bernalillo    NEW MEXICO            11.14                           25.69   
4     Broward       FLORIDA            23.07                           30.13   

   min_of_monthly_avg_temperature  
0                          -11.42  
1                          -12.44  
2                            1.67  
3                           -5.07  
4                           12.96

### Step 3: Define the Data Model
#### 3.1 Conceptual Data Model
#### Tables:
##### Fact Table:  
###### 1. immigration
 - description: contains i94 immigrations data
 - columns: *id, created_date, year, month, admission_no, admitted_date, departure_date, arrival_date, port_name, airport_name, entry_state_id, entry_state_name, entry_county, airline, flight_no, gender, birth_year, age, origin_country_name, resident_country_name, visa_type, visa_type_name, state_code, current_state_name*
 
##### Dimension Tables:  

###### 2. demographic
 - description:  contains demographic information by US cities
 - columns:  *State(Primary key), State Code, City, Count, Race, Median Age, Male Population, Female Population, Number of Veterans, Foreign-born, Average Household Size*
 
###### 3. crime
 - description: contains crime information in different categories by US states
 - columns:  *county(Composite key), state(Composite key), category, total_crime, Violent_crime, Murder_and_nonnegligent_manslaughter, Robbery, Aggravated_assault, Property_crime, Burglary, Larceny-theft, Motor_vehicle_theft, Arson3, Rape, crime_rate, violent_crime_rate, property_crime_rate*
 
###### 4. temperature
 - description:  contains temperatures by US counties
 - columns:  *county(Composite key), state(Composite key), avg_temperature, max_of_monthly_avg_temperature, min_of_monthly_avg_temperature*
 
###### 5. date
 - description:  contains date information related to Immigration table
 - columns:  *created_date(Primary key), day, week, month, year, weekday*

##### Why I chose this model:
Along with Immigration table which contains key information about each immigrant in the US, I provided with a variety of dimension tables that could be useful for exploring immigration table more in details.
To do so, I have transformed all tables in a way that one can join them with immigration table by cleaning fact table to make foreign keys and by setting primary keys in each dimension table.


#### 3.2 Mapping Out Data Pipelines
List the steps necessary to pipeline the data into the chosen data model

In [18]:
%run 'create_tables.py'

In [19]:
# Define connecion and curser
conn = psycopg2.connect("host=127.0.0.1 dbname=immigrationdb user=student password=student")
cur = conn.cursor()

In [20]:
# Convert pyspark tables to pandas dataframes
immigration = immigration.toPandas()
date = date.toPandas()

# Insert data into created tables
for index, row in demographic.iterrows():
    cur.execute(demographic_table_insert, list(row.values))
    conn.commit()

for index, row in immigration.iterrows():
    cur.execute(immigration_table_insert, list(row.values))
    conn.commit()
    
for index, row in crime.iterrows():
    cur.execute(crime_table_insert, list(row.values))
    conn.commit()
    
for index, row in date.iterrows():
    cur.execute(date_table_insert, list(row.values))
    conn.commit()
    
for index, row in temperature.iterrows():
    cur.execute(temperature_table_insert, list(row.values))
    conn.commit()

### Step 4: Run Pipelines to Model the Data 
#### 4.1 Create the data model
1. Create database and tables by running *'create_tables.py'* file
2. Extract, transform, and load data by running *'etl.py'* file

#### 4.2 Data Quality Checks
Explain the data quality checks you'll perform to ensure the pipeline ran as expected. These could include:
 * Integrity constraints on the relational database (e.g., unique key, data type, etc.)
 * Unit tests for the scripts to ensure they are doing the right thing
 * Source/Count checks to ensure completeness
 
Run Quality Checks

In [21]:
def data_check(cur, conn):
    """ 
    This data checks tables which data was inserted into to make sure they contain values,
    and raising error messages if they have no results or cantain no rows.
    """
    failing_tests = []
    for key in data_check_queries:
        cur.execute(data_check_queries[key])
        sql_result = cur.fetchone()
        
        exp_result = 0
        
        error_count = 0
        
        if exp_result != sql_result [0]:
            error_count += 1
            failing_tests.append(data_check_queries[key])
            
        if error_count > 0:
            print(f'{key}: SQL Tests Failed')
            raise ValueError('Data quality check failed')
            
        if error_count == 0:
            print(f'{key}: SQL Tests Passed') 
            
data_check(cur, conn)

immigration_table_check: SQL Tests Passed
demographic_table_check: SQL Tests Passed
crime_table_check: SQL Tests Passed
date_table_check: SQL Tests Passed
temperature_table_check: SQL Tests Passed


In [22]:
demographic.shape

(49, 14)

In [23]:
demographic.isnull().sum(axis = 0)

state                                         0
state_code                                    0
population                                    0
median_age                                    0
male_population                               0
female_population                             0
number_of_veterans                            0
foreign_born                                  0
average_household_size                        0
AmericanIndian_and_AlaskaNative_population    0
Asian_population                              0
Black_or_AfricanAmerican_population           0
Hispanic_or_Latino_population                 0
White_population                              0
dtype: int64

In [29]:
%load_ext sql
%sql postgresql://student:student@127.0.0.1/immigrationdb

%sql SELECT \
sum(case when state is null then 1 else 0 end) as state_null_count, \
sum(case when state_code is null then 1 else 0 end) as state_code_null_count,\
sum(case when population is null then 1 else 0 end) as population_null_count, \
sum(case when median_age is null then 1 else 0 end) as median_age_null_count, \
sum(case when male_population is null then 1 else 0 end) as male_population_null_count, \
sum(case when female_population is null then 1 else 0 end) as female_population_null_count, \
sum(case when number_of_veterans is null then 1 else 0 end) as number_of_veterans_null_count, \
sum(case when foreign_born is null then 1 else 0 end) as foreign_born_null_count, \
sum(case when average_household_size is null then 1 else 0 end) as average_household_size_null_count, \
sum(case when AmericanIndian_and_AlaskaNative_population is null then 1 else 0 end) as AmericanIndian_and_AlaskaNative_population_null_count, \
sum(case when Asian_population is null then 1 else 0 end) as Asian_population_null_count, \
sum(case when Black_or_AfricanAmerican_population is null then 1 else 0 end) as Black_or_AfricanAmerican_population_null_count, \
sum(case when Hispanic_or_Latino_population is null then 1 else 0 end) as Hispanic_or_Latino_population_null_count, \
sum(case when White_population is null then 1 else 0 end) as White_population_null_count \
from demographic;

The sql extension is already loaded. To reload it, use:
  %reload_ext sql
 * postgresql://student:***@127.0.0.1/immigrationdb
1 rows affected.


[(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0)]

In [30]:
%sql SELECT * FROM demographic LIMIT 5;

 * postgresql://student:***@127.0.0.1/immigrationdb
5 rows affected.


[('ALABAMA', 'AL', 1096154, 38.0, 2448200, 2715106, 352896, 252541, 2.0, 8084, 28769, 521068, 39313, 498920),
 ('ALASKA', 'AK', 336228, 32.2, 764725, 728750, 137460, 166290, 2.0, 36339, 36825, 23107, 27261, 212696),
 ('ARIZONA', 'AZ', 5754881, 34.1, 11137275, 11360435, 1322525, 3411565, 2.19, 129708, 229183, 296222, 1508157, 3591611),
 ('ARKANSAS', 'AR', 643597, 32.6, 1400724, 1482165, 154390, 307753, 2.17, 9381, 22062, 149608, 77813, 384733),
 ('CALIFORNIA', 'CA', 31753718, 35.8, 61055672, 62388681, 4617022, 37059662, 2.59, 401386, 4543730, 2047009, 9856464, 14905129)]

In [31]:
immigration.shape

(23607, 22)

In [32]:
immigration.isnull().sum(axis = 0)

id                           0
created_date                 0
admission_no                 0
admitted_date                0
departure_date               0
arrival_date                 0
port_name                    0
airport_name             12370
entry_state_id           12370
entry_state_name         12370
entry_county             12370
airline                      0
flight_no                    0
gender                       0
birth_year                   0
age                          0
origin_country_name       4340
resident_country_name     1552
visa_type                    0
visa_type_name               0
state_code                   0
current_state_name         990
dtype: int64

In [33]:
%sql SELECT count(*) FROM immigration;

 * postgresql://student:***@127.0.0.1/immigrationdb
1 rows affected.


[(23607,)]

In [34]:
%sql SELECT \
sum(case when id is null then 1 else 0 end) as id, \
sum(case when created_date is null then 1 else 0 end) as created_date,\
sum(case when admission_no is null then 1 else 0 end) as admission_no, \
sum(case when admitted_date is null then 1 else 0 end) as admitted_date, \
sum(case when departure_date is null then 1 else 0 end) as departure_date, \
sum(case when arrival_date is null then 1 else 0 end) as arrival_date, \
sum(case when port_name is null then 1 else 0 end) as port_name, \
sum(case when airport_name is null then 1 else 0 end) as airport_name, \
sum(case when entry_state_id is null then 1 else 0 end) as entry_state_id, \
sum(case when entry_state_name is null then 1 else 0 end) as entry_state_name, \
sum(case when entry_county is null then 1 else 0 end) as entry_county, \
sum(case when airline is null then 1 else 0 end) as airline, \
sum(case when flight_no is null then 1 else 0 end) as flight_no, \
sum(case when birth_year is null then 1 else 0 end) as birth_year, \
sum(case when age is null then 1 else 0 end) as age, \
sum(case when origin_country_name is null then 1 else 0 end) as origin_country_name, \
sum(case when resident_country_name is null then 1 else 0 end) as resident_country_name, \
sum(case when visa_type is null then 1 else 0 end) as visa_type, \
sum(case when visa_type_name is null then 1 else 0 end) as visa_type_name, \
sum(case when state_code is null then 1 else 0 end) as state_code, \
sum(case when current_state_name is null then 1 else 0 end) as current_state_name \
from immigration;

 * postgresql://student:***@127.0.0.1/immigrationdb
1 rows affected.


[(0, 0, 0, 0, 0, 0, 0, 12370, 12370, 12370, 12370, 0, 0, 0, 0, 4340, 1552, 0, 0, 0, 990)]

In [35]:
%sql SELECT * FROM immigration LIMIT 5;

 * postgresql://student:***@127.0.0.1/immigrationdb
5 rows affected.


[(1973857, datetime.date(2016, 4, 11), 2147483647, datetime.date(2016, 7, 9), datetime.date(2016, 4, 25), datetime.date(2016, 4, 11), 'NEWARK/TETERBORO', 'Lakefront Airport', 'LA', 'LOUISIANA', 'New Orleans', 'UA', '00998', 'F', 1990, 26, 'BELGIUM', 'BELGIUM', 2, 'Pleasure', 'AZ', 'ARIZONA'),
 (3700217, datetime.date(2016, 4, 20), 2147483647, datetime.date(2016, 10, 19), datetime.date(2016, 5, 22), datetime.date(2016, 4, 20), 'NEWARK/TETERBORO', 'Lakefront Airport', 'LA', 'LOUISIANA', 'New Orleans', 'UA', '00961', 'M', 1981, 35, 'INDIA', 'INDIA', 1, 'Business', 'SC', 'SOUTH CAROLINA'),
 (641170, datetime.date(2016, 4, 4), 2147483647, datetime.date(2016, 7, 2), datetime.date(2016, 4, 7), datetime.date(2016, 4, 4), 'NEWARK/TETERBORO', 'Lakefront Airport', 'LA', 'LOUISIANA', 'New Orleans', 'LH', '00402', 'M', 1966, 50, 'ITALY', 'ITALY', 1, 'Business', 'SC', 'SOUTH CAROLINA'),
 (4280680, datetime.date(2016, 4, 23), 2147483647, datetime.date(2016, 7, 21), datetime.date(2016, 4, 30), datetime.date(2016, 4, 23), 'NEWARK/TETERBORO', 'Lakefront Airport', 'LA', 'LOUISIANA', 'New Orleans', 'BA', '00185', 'M', 1958, 58, 'SWITZERLAND', 'SWITZERLAND', 2, 'Pleasure', 'SC', 'SOUTH CAROLINA'),
 (3892965, datetime.date(2016, 4, 21), 2147483647, datetime.date(2016, 10, 20), datetime.date(2016, 9, 10), datetime.date(2016, 4, 21), 'NEWARK/TETERBORO', 'Lakefront Airport', 'LA', 'LOUISIANA', 'New Orleans', 'UA', '00180', 'F', 1960, 56, 'CHINA', 'CHINA', 2, 'Pleasure', 'SC', 'SOUTH CAROLINA')]

In [36]:
crime.shape

(2577, 17)

In [37]:
crime.isnull().sum(axis = 0)

county                                  0
state                                   0
category                                0
total_crime                             0
violent_crime                           0
murder_and_nonnegligent_manslaughter    0
robbery                                 0
aggravated_assault                      0
property_crime                          0
burglary                                0
larceny-theft                           0
motor_vehicle_theft                     0
arson3                                  0
rape                                    0
crime_rate                              0
violent_crime_rate                      0
property_crime_rate                     0
dtype: int64

In [38]:
%sql SELECT count(*) FROM crime;

 * postgresql://student:***@127.0.0.1/immigrationdb
1 rows affected.


[(2555,)]

In [39]:
%sql SELECT \
sum(case when county is null then 1 else 0 end) as county, \
sum(case when state is null then 1 else 0 end) as state,\
sum(case when category is null then 1 else 0 end) as category,\
sum(case when total_crime is null then 1 else 0 end) as total_crime, \
sum(case when violent_crime is null then 1 else 0 end) as violent_crime, \
sum(case when murder_and_nonnegligent_manslaughter is null then 1 else 0 end) as murder_and_nonnegligent_manslaughter, \
sum(case when robbery is null then 1 else 0 end) as robbery, \
sum(case when aggravated_assault is null then 1 else 0 end) as aggravated_assault, \
sum(case when property_crime is null then 1 else 0 end) as property_crime, \
sum(case when burglary is null then 1 else 0 end) as burglary, \
sum(case when larceny_theft is null then 1 else 0 end) as larceny_theft, \
sum(case when motor_vehicle_theft is null then 1 else 0 end) as motor_vehicle_theft, \
sum(case when arson3 is null then 1 else 0 end) as arson3, \
sum(case when rape is null then 1 else 0 end) as rape, \
sum(case when crime_rate is null then 1 else 0 end) as crime_rate, \
sum(case when violent_crime_rate is null then 1 else 0 end) as violent_crime_rate, \
sum(case when property_crime_rate is null then 1 else 0 end) as property_crime_rate \
from crime;

 * postgresql://student:***@127.0.0.1/immigrationdb
1 rows affected.


[(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0)]

In [40]:
%sql SELECT * FROM crime LIMIT 5;

 * postgresql://student:***@127.0.0.1/immigrationdb
5 rows affected.


[('Autauga', 'ALABAMA', ' Metropolitan Counties', 419, 69, 0, 6, 50, 344, 111, 187, 46, 0, 13, 972.5, 160.1, 798.4),
 ('Baldwin', 'ALABAMA', ' Metropolitan Counties', 796, 115, 0, 33, 73, 648, 225, 390, 33, 0, 9, 406.1, 58.7, 330.6),
 ('Bibb', 'ALABAMA', ' Metropolitan Counties', 49, 7, 0, 1, 4, 41, 20, 18, 3, 0, 2, 366.6, 52.4, 306.8),
 ('Blount', 'ALABAMA', ' Metropolitan Counties', 1046, 204, 5, 5, 178, 832, 247, 503, 82, 0, 16, 5878.7, 1146.5, 4676.0),
 ('Calhoun', 'ALABAMA', ' Metropolitan Counties', 430, 16, 0, 1, 11, 413, 181, 225, 7, 0, 4, 309.1, 11.5, 296.9)]

In [41]:
temperature.shape

(208, 5)

In [42]:
temperature.isnull().sum(axis = 0)

county                            0
state                             0
avg_temperature                   0
max_of_monthly_avg_temperature    0
min_of_monthly_avg_temperature    0
dtype: int64

In [43]:
%sql SELECT count(*) FROM temperature;

 * postgresql://student:***@127.0.0.1/immigrationdb
1 rows affected.


[(208,)]

In [44]:
%sql SELECT \
sum(case when county is null then 1 else 0 end) as county, \
sum(case when state is null then 1 else 0 end) as state,\
sum(case when avg_temperature is null then 1 else 0 end) as avg_temperature, \
sum(case when max_of_monthly_avg_temperature is null then 1 else 0 end) as max_of_monthly_avg_temperature, \
sum(case when min_of_monthly_avg_temperature is null then 1 else 0 end) as min_of_monthly_avg_temperature \
from temperature;

 * postgresql://student:***@127.0.0.1/immigrationdb
1 rows affected.


[(0, 0, 0, 0, 0)]

In [45]:
%sql SELECT * FROM temperature LIMIT 5;

 * postgresql://student:***@127.0.0.1/immigrationdb
5 rows affected.


[('Allegheny', 'PENNSYLVANIA', 9.61, 28.6, -11.42),
 ('Allen', 'INDIANA', 9.7, 29.04, -12.44),
 ('Bell', 'TEXAS', 18.54, 32.66, 1.67),
 ('Bernalillo', 'NEW MEXICO', 11.14, 25.69, -5.07),
 ('Broward', 'FLORIDA', 23.07, 30.13, 12.96)]

In [46]:
date.shape

(71, 6)

In [47]:
date.isnull().sum(axis = 0)

created_date    0
day             0
week            0
month           0
year            0
weekday         0
dtype: int64

In [48]:
%sql SELECT \
sum(case when created_date is null then 1 else 0 end) as created_date, \
sum(case when day is null then 1 else 0 end) as day,\
sum(case when week is null then 1 else 0 end) as week, \
sum(case when month is null then 1 else 0 end) as month, \
sum(case when year is null then 1 else 0 end) as year, \
sum(case when weekday is null then 1 else 0 end) as weekday \
from date;

 * postgresql://student:***@127.0.0.1/immigrationdb
1 rows affected.


[(0, 0, 0, 0, 0, 0)]

In [49]:
%sql SELECT count(*) FROM date;

 * postgresql://student:***@127.0.0.1/immigrationdb
1 rows affected.


[(71,)]

In [50]:
%sql SELECT * FROM date LIMIT 5;

 * postgresql://student:***@127.0.0.1/immigrationdb
5 rows affected.


[(datetime.date(2016, 6, 23), 23, 25, 6, 2016, 5),
 (datetime.date(2016, 4, 14), 14, 15, 4, 2016, 5),
 (datetime.date(2016, 5, 24), 24, 21, 5, 2016, 3),
 (datetime.date(2016, 5, 26), 26, 21, 5, 2016, 5),
 (datetime.date(2016, 5, 3), 3, 18, 5, 2016, 3)]

#### 4.3 Data dictionary 

###### 1. immigration
- This table contains i94 immigrations data.
- It comes from "US National Tourism and Trade Office".
- It shows you information about each immigrant id whose resident country was Japan, and who was registered in i94 system during 2016.

| Field Name | Data Type | Description | Example |
| --- | --- | --- | --- |
| **id** *Primary key* | int | unique id of I94 immigration record | 4526055 |
| created_date *Foreign key* | date | date the data was created | 4/24/2016 |
| admission_no | int | admission no. of immigrants | 2147483647 |
| admitted_date | date | date the immigrants were admitted for the entry | 7/22/2016 |
| departure_date | date | date the immigrants departed the US | 4/28/2016 |
| arrival_date | date | date the immigrants arrived to the US | 4/24/2016 |
| port_name | varchar | airport code of immigrants's entry | NEWARK/TETERBORO |
| airport_name | varchar | airport name of immigrants's entry | Lakefront Airport |
| entry_state_id | varchar | state code of immigrants' entry | LA |
| **entry_state_name**　*Foreign key* | varchar | state name of immigrants' entry | LOUISIANA |
| **entry_county**　*Foreign key* | varchar | county name of immigrants' entry | New Orleans |
| airline | varchar | airline used by immigrants for entry | JL |
| flight_no | varchar | flight no. used by immigrants for entry | 60 |
| gender | varchar | gender of immigrants | M |
| birth_year | int | birth year of immigrants | 1971 |
| age | int | age of immigrants | 45 |
| origin_country_name | varchar | original country of immigrants | JAPAN |
| resident_country_name | varchar | resident country of immigrants | JAPAN |
| visa_type | bigint | immigrants' visa type code | 1 |
| visa_type_name | varchar | immigrants' visa type name | Business |
| state_code | varchar | state code of immigrants's final destination | AZ |
| **current_state_name**　*Foreign key* | varchar | state name of immigrants' final destination | ARIZONA |

###### 2. demographic
- This table contains basic demographic information of the US of 2015.
- It comes from "OpenSoft".
- It allows you to join by Primary key 'state' with the fact table to dig in more deeply about the satistics of the places immigrants chose to enter or end up staying in the end.

| Field Name | Data Type | Description | Example |
| --- | --- | --- | --- |
| **state** *Primary key* | varchar | state name | ALABAMA |
| state_code | varchar | state code | AL |
| population | int | population | 1096154 |
| median_age | float | median age | 38 |
| male_population | int | male population | 2448200 |
| female_population | int | female population | 2715106 |
| number_of_veterans | int | number of veterans | 352896 |
| foreign_born | int | number of foreign-born | 252541 |
| average_household_size | float | avg of household size | 2 |
| AmericanIndian_and_AlaskaNative_population | int | Native American and Alaskan Native population | 8084 |
| Asian_population | int | Asian population | 28769 |
| Black_or_AfricanAmerican_population | int | Black of African Americn population | 521068 | 
| Hispanic_or_Latino_population | int | Hispanic or Latino population | 39313 |
| White_population | int | White population| 498920 |

###### 3. crime
- This table contains number of crimes by US state and county of 2015.
- It comes from "data.world".
- It allows you to join by Composite key 'state' and 'county' with the fact table to understand crime situations including crime rate of the places immigrants chose to enter or end up staying in the end.

| Field Name | Data Type | Description | Example |
| --- | --- | --- | --- |
| **county** *Composite key* | varchar | county name | Autauga |
| **state** *Composite key* | varchar | state name | ALABAMA |
| category | varchar | category of county | Metropolitan Counties |
| total_crime | int | total number of crimes | 2/22/1901 |
| violent_crime | int | number of violent crimes | 3/9/1900 |
| murder_and_nonnegligent_manslaughter | int | number of murders and nonnegligent manslaughters | 1/0/1900 |
| robbery | int | number of robberies | 6 |
| aggravated_assault | int | number of aggravated assaults | 50 |
| burglary | int | number of burglaries | 344 |
| property_crime | int | number of property crimes | 111 |
| larceny_theft | int | number of larceny thefts | 187 |
| motor_vehicle_theft | int | number of motor vehicle thefts | 46 |
| arson3 | int | number of arsons | 0 |
| rape | int | number of rapes | 13 |
| crime_rate | float | crime rate(per 100,000 people) | 972.5 |
| violent_crime_rate | float | crime rate of violent crimes(per 100,000 people) | 160.1 |
| property_crime_rate | float | crime rate of property crimes(per 100,000 people) | 798.4 |

###### 4. temperature
- This table contains global temperature information since 1970s by US state and county.
- It comes from "Kaggle".
- It allows you to join by Composite key 'state' and 'county' with the fact table to understand the statistics of temperature of the places immigrants chose to enter or end up staying in the end.

| Field Name | Data Type | Description | Example |
| --- | --- | --- | --- |
| **county** *Composite key* | varchar | county name | Allegheny |
| **state** *Composite key* | varchar | state name | PENNSYLVANIA |
| avg_temperature | float | average of monthly average temperature in 1970-2016  | 9.61 |
| max_of_monthly_avg_temperature | float | max value of of monthly average temperature in 1970-2016 | 28.6 |
| min_of_monthly_avg_temperature | float | minimum value of of monthly average temperature in 1970-2016 | -11.42 |

###### 5. date
- This table is the breakdown of immigration table's 'created date' column.
- It allows you to join by Primary key 'created_date' with the fact table to analyze from date perspective.

| Field Name | Data Type | Description | Example |
| --- | --- | --- | --- |
| **created_date** *Primary key* | date | date the immigration record was created | 6/23/2016 |
| day | int | date of created_date | 23 |
| week | int | week of created_date | 25 |
| month | int | month of created_date | 6 |
| year | int | year of created_date | 2016 |
| weekday | int | weekday of created_date | 5 |

#### Step 5: Complete Project Write Up
###### 1. Why I chose the technologies I used for this project
- `Jupyter Notebook` - I used notebook to better understand the data as I process them visually.
- `Python` - Python is used for creating the actual scripts to execute the whole process because it is one of the simplest languages to work with.
- `Pandas` - I used pandas because pandas is convinient for checking, cleaning, merging and visualizing the results after each manipulation process.
- `PySpark` - I used PySpark to work with large data, i94 immigration data, making use of distributed processing for more efficiency. I did not create a EMR cluster to use PySpark for this project since the project workspace did not recquired an EMR cluster, but if you do it locally you would need to create one.
- `PostgreSQL` - I used PostgreSQL to create a database and tables because it is recommended to use for relational models, and also because I'm familiar with it.

###### 2. How often this data should be updated and why

* Monthly update is recommended since the original i94 data itself is updated on monthly basis.

###### 3. How I would approach the problem differently under the following scenarios:

* *The data was increased by 100x.*
  *  I would use PySpark instead of pandas for all the data, and for writing into tables I would still use PostgreSQL, but suggest upgrading a hardware(i.e. better CPU, RAM, and SSD) to execute. If I wanted to use cloud infrustructure, I would use AWS S3 to write and load into those tables.
* *The data populates a dashboard that must be updated on a daily basis by 7am every day.*
  *  I would implement Apache Airflow to create a DAG file that executes ETL pipelines to run everyday before 7 am.
* *The database needed to be accessed by 100+ people.*
  *  I would probably stick to using PostgreSQL as a database, however, if it is too slow I would move them to cloud data warehouse such as Redshift on AWS.